File: app/src/main/java/fcu/app/appclassfinalproject/helper/ChatHelper.java
package fcu.app.appclassfinalproject.helper;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;
import fcu.app.appclassfinalproject.model.ChatRoom;
import fcu.app.appclassfinalproject.model.Message;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class ChatHelper {

  private static final String TAG = "ChatHelper";

  /**
   * 建立一對一聊天室
   */
  public static int createPrivateChatRoom(SQLiteDatabase db, int userId1, int userId2) {
    try {
      // 檢查是否已存在聊天室
      String checkQuery = "SELECT cr.id FROM ChatRooms cr " +
          "INNER JOIN ChatRoomMembers crm1 ON cr.id = crm1.chatroom_id " +
          "INNER JOIN ChatRoomMembers crm2 ON cr.id = crm2.chatroom_id " +
          "WHERE cr.type = 'private' " +
          "AND crm1.user_id = ? AND crm2.user_id = ? " +
          "AND crm1.user_id != crm2.user_id";

      Cursor cursor = db.rawQuery(checkQuery, new String[]{
          String.valueOf(userId1), String.valueOf(userId2)
      });

      if (cursor.moveToFirst()) {
        int existingRoomId = cursor.getInt(0);
        cursor.close();
        Log.d(TAG, "找到現有的一對一聊天室: " + existingRoomId);
        return existingRoomId;
      }
      cursor.close();

      // 建立新聊天室
      String currentTime = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
          .format(new Date());

      ContentValues roomValues = new ContentValues();
      roomValues.put("name", null); // 一對一聊天室沒有名稱
      roomValues.put("type", "private");
      roomValues.put("created_at", currentTime);
      roomValues.put("created_by", userId1);

      long roomId = db.insert("ChatRooms", null, roomValues);

      if (roomId != -1) {
        // 添加兩個成員
        addMemberToChatRoom(db, (int) roomId, userId1);
        addMemberToChatRoom(db, (int) roomId, userId2);
        Log.d(TAG, "建立一對一聊天室成功: " + roomId);
        return (int) roomId;
      }

    } catch (Exception e) {
      Log.e(TAG, "建立一對一聊天室失敗: " + e.getMessage(), e);
    }
    return -1;
  }

  /**
   * 建立群組聊天室
   */
  public static int createGroupChatRoom(SQLiteDatabase db, String groupName, int creatorId,
      List<Integer> memberIds) {
    try {
      String currentTime = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
          .format(new Date());

      ContentValues roomValues = new ContentValues();
      roomValues.put("name", groupName);
      roomValues.put("type", "group");
      roomValues.put("created_at", currentTime);
      roomValues.put("created_by", creatorId);

      long roomId = db.insert("ChatRooms", null, roomValues);

      if (roomId != -1) {
        // 添加所有成員
        for (Integer memberId : memberIds) {
          addMemberToChatRoom(db, (int) roomId, memberId);
        }
        // 添加創建者
        addMemberToChatRoom(db, (int) roomId, creatorId);
        Log.d(TAG, "建立群組聊天室成功: " + roomId);
        return (int) roomId;
      }

    } catch (Exception e) {
      Log.e(TAG, "建立群組聊天室失敗: " + e.getMessage(), e);
    }
    return -1;
  }

  /**
   * 添加成員到聊天室
   */
  private static void addMemberToChatRoom(SQLiteDatabase db, int chatroomId, int userId) {
    try {
      String currentTime = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
          .format(new Date());

      ContentValues values = new ContentValues();
      values.put("chatroom_id", chatroomId);
      values.put("user_id", userId);
      values.put("joined_at", currentTime);

      db.insertWithOnConflict("ChatRoomMembers", null, values,
          SQLiteDatabase.CONFLICT_IGNORE);
    } catch (Exception e) {
      Log.e(TAG, "添加成員失敗: " + e.getMessage(), e);
    }
  }

  /**
   * 獲取用戶的所有聊天室
   */
  public static List<ChatRoom> getChatRoomsByUser(SQLiteDatabase db, int userId) {
    List<ChatRoom> chatRooms = new ArrayList<>();
    try {
      String query = "SELECT DISTINCT cr.id, cr.name, cr.type, cr.created_at, cr.created_by " +
          "FROM ChatRooms cr " +
          "INNER JOIN ChatRoomMembers crm ON cr.id = crm.chatroom_id " +
          "WHERE crm.user_id = ? " +
          "ORDER BY cr.created_at DESC";

      Cursor cursor = db.rawQuery(query, new String[]{String.valueOf(userId)});

      while (cursor.moveToNext()) {
        int id = cursor.getInt(0);
        String name = cursor.isNull(1) ? null : cursor.getString(1);
        String type = cursor.getString(2);
        String createdAt = cursor.getString(3);
        int createdBy = cursor.getInt(4);

        ChatRoom chatRoom = new ChatRoom(id, name, type, createdAt, createdBy);

        // 獲取成員名稱列表
        List<String> memberNames = getChatRoomMemberNames(db, id, userId);
        chatRoom.setMemberNames(memberNames);

        // 獲取最後一條訊息
        Message lastMessage = getLastMessage(db, id);
        if (lastMessage != null) {
          chatRoom.setLastMessage(lastMessage.getContent());
          chatRoom.setLastMessageTime(lastMessage.getCreatedAt());
        }

        chatRooms.add(chatRoom);
      }
      cursor.close();

    } catch (Exception e) {
      Log.e(TAG, "獲取聊天室列表失敗: " + e.getMessage(), e);
    }
    return chatRooms;
  }

  /**
   * 獲取聊天室成員名稱（排除當前用戶）
   */
  private static List<String> getChatRoomMemberNames(SQLiteDatabase db, int chatroomId,
      int currentUserId) {
    List<String> memberNames = new ArrayList<>();
    try {
      String query = "SELECT u.account FROM Users u " +
          "INNER JOIN ChatRoomMembers crm ON u.id = crm.user_id " +
          "WHERE crm.chatroom_id = ? AND u.id != ?";

      Cursor cursor = db.rawQuery(query,
          new String[]{String.valueOf(chatroomId), String.valueOf(currentUserId)});

      while (cursor.moveToNext()) {
        memberNames.add(cursor.getString(0));
      }
      cursor.close();

    } catch (Exception e) {
      Log.e(TAG, "獲取成員名稱失敗: " + e.getMessage(), e);
    }
    return memberNames;
  }

  /**
   * 獲取聊天室的最後一條訊息
   */
  private static Message getLastMessage(SQLiteDatabase db, int chatroomId) {
    try {
      String query = "SELECT id, sender_id, content, created_at " +
          "FROM Messages " +
          "WHERE chatroom_id = ? " +
          "ORDER BY created_at DESC " +
          "LIMIT 1";

      Cursor cursor = db.rawQuery(query, new String[]{String.valueOf(chatroomId)});

      if (cursor.moveToFirst()) {
        int id = cursor.getInt(0);
        int senderId = cursor.getInt(1);
        String content = cursor.getString(2);
        String createdAt = cursor.getString(3);

        cursor.close();
        return new Message(id, chatroomId, senderId, content, createdAt);
      }
      cursor.close();

    } catch (Exception e) {
      Log.e(TAG, "獲取最後訊息失敗: " + e.getMessage(), e);
    }
    return null;
  }

  /**
   * 發送訊息
   */
  public static long sendMessage(SQLiteDatabase db, int chatroomId, int senderId,
      String content) {
    try {
      String currentTime = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
          .format(new Date());

      ContentValues values = new ContentValues();
      values.put("chatroom_id", chatroomId);
      values.put("sender_id", senderId);
      values.put("content", content);
      values.put("created_at", currentTime);

      long messageId = db.insert("Messages", null, values);
      Log.d(TAG, "發送訊息成功: " + messageId);
      return messageId;

    } catch (Exception e) {
      Log.e(TAG, "發送訊息失敗: " + e.getMessage(), e);
      return -1;
    }
  }

  /**
   * 獲取聊天室的所有訊息
   */
  public static List<Message> getMessagesByChatRoom(SQLiteDatabase db, int chatroomId,
      int currentUserId) {
    List<Message> messages = new ArrayList<>();
    try {
      String query = "SELECT m.id, m.sender_id, m.content, m.created_at, u.account " +
          "FROM Messages m " +
          "INNER JOIN Users u ON m.sender_id = u.id " +
          "WHERE m.chatroom_id = ? " +
          "ORDER BY m.created_at ASC";

      Cursor cursor = db.rawQuery(query, new String[]{String.valueOf(chatroomId)});

      while (cursor.moveToNext()) {
        int id = cursor.getInt(0);
        int senderId = cursor.getInt(1);
        String content = cursor.getString(2);
        String createdAt = cursor.getString(3);
        String senderName = cursor.getString(4);

        Message message = new Message(id, chatroomId, senderId, content, createdAt);
        message.setSenderName(senderName);
        message.setSentByMe(senderId == currentUserId);

        messages.add(message);
      }
      cursor.close();

    } catch (Exception e) {
      Log.e(TAG, "獲取訊息列表失敗: " + e.getMessage(), e);
    }
    return messages;
  }
}

File: app/src/main/java/fcu/app/appclassfinalproject/supabase/ChatHelper.kt
package fcu.app.appclassfinalproject.supabase

import fcu.app.appclassfinalproject.model.ChatRoom
import fcu.app.appclassfinalproject.model.Message
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.gotrue.auth
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.Serializable
import java.text.SimpleDateFormat
import java.util.*

@Serializable
data class ChatRoomData(
    val id: Int? = null,
    val name: String? = null,
    val type: String,
    val created_at: String,
    val created_by: String
)

@Serializable
data class ChatRoomMemberData(
    val id: Int? = null,
    val chatroom_id: Int,
    val user_id: String,
    val joined_at: String
)

@Serializable
data class MessageData(
    val id: Int? = null,
    val chatroom_id: Int,
    val sender_id: String,
    val content: String,
    val created_at: String
)

@Serializable
data class UserData(
    val id: String,
    val account: String? = null,
    val email: String
)

object ChatHelper {

    private val client = SupabaseManager.getClient()
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())

    /**
     * 建立一對一聊天室
     */
    fun createPrivateChatRoom(userId1: String, userId2: String): Int? {
        return runBlocking {
            try {
                // 檢查是否已存在聊天室
                val existingRooms = client.from("chatrooms")
                    .select {
                        filter {
                            eq("type", "private")
                        }
                    }
                    .decodeList<ChatRoomData>()

                // 檢查兩個用戶是否已經在同一個聊天室
                for (room in existingRooms) {
                    val members = client.from("chatroom_members")
                        .select {
                            filter {
                                eq("chatroom_id", room.id!!)
                            }
                        }
                        .decodeList<ChatRoomMemberData>()

                    val memberIds = members.map { it.user_id }.toSet()
                    if (memberIds.contains(userId1) && memberIds.contains(userId2)) {
                        return@runBlocking room.id
                    }
                }

                // 建立新聊天室
                val newRoom = ChatRoomData(
                    name = null,
                    type = "private",
                    created_at = dateFormat.format(Date()),
                    created_by = userId1
                )

                val createdRoom = client.from("chatrooms")
                    .insert(newRoom) {
                        select(Columns.ALL)
                    }
                    .decodeSingle<ChatRoomData>()

                // 添加兩個成員
                val member1 = ChatRoomMemberData(
                    chatroom_id = createdRoom.id!!,
                    user_id = userId1,
                    joined_at = dateFormat.format(Date())
                )
                val member2 = ChatRoomMemberData(
                    chatroom_id = createdRoom.id!!,
                    user_id = userId2,
                    joined_at = dateFormat.format(Date())
                )

                client.from("chatroom_members").insert(member1)
                client.from("chatroom_members").insert(member2)

                createdRoom.id
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    /**
     * 建立群組聊天室
     */
    fun createGroupChatRoom(groupName: String, creatorId: String, memberIds: List<String>): Int? {
        return runBlocking {
            try {
                val newRoom = ChatRoomData(
                    name = groupName,
                    type = "group",
                    created_at = dateFormat.format(Date()),
                    created_by = creatorId
                )

                val createdRoom = client.from("chatrooms")
                    .insert(newRoom) {
                        select(Columns.ALL)
                    }
                    .decodeSingle<ChatRoomData>()

                // 添加所有成員
                val now = dateFormat.format(Date())
                val members = (memberIds + creatorId).map { userId ->
                    ChatRoomMemberData(
                        chatroom_id = createdRoom.id!!,
                        user_id = userId,
                        joined_at = now
                    )
                }

                client.from("chatroom_members").insert(members)

                createdRoom.id
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    /**
     * 獲取用戶的所有聊天室
     */
    fun getChatRoomsByUser(userId: String): List<ChatRoom> {
        return runBlocking {
            try {
                // 獲取用戶參與的聊天室 ID
                val memberRooms = client.from("chatroom_members")
                    .select {
                        filter {
                            eq("user_id", userId)
                        }
                    }
                    .decodeList<ChatRoomMemberData>()

                val roomIds = memberRooms.map { it.chatroom_id }.distinct()

                if (roomIds.isEmpty()) {
                    return@runBlocking emptyList()
                }

                // 獲取聊天室資訊（簡化版本：逐個查詢）
                val rooms = roomIds.flatMap { roomId ->
                    try {
                        client.from("chatrooms")
                            .select {
                                filter {
                                    eq("id", roomId)
                                }
                            }
                            .decodeList<ChatRoomData>()
                    } catch (e: Exception) {
                        emptyList()
                    }
                }.sortedByDescending { it.created_at }

                // 轉換為 ChatRoom 模型
                rooms.map { roomData ->
                    val chatRoom = ChatRoom(
                        roomData.id!!,
                        roomData.name ?: "",
                        roomData.type,
                        roomData.created_at,
                        roomData.created_by.toIntOrNull() ?: 0
                    )

                    // 獲取成員名稱
                    val members = client.from("chatroom_members")
                        .select {
                            filter {
                                eq("chatroom_id", roomData.id!!)
                                neq("user_id", userId)
                            }
                        }
                        .decodeList<ChatRoomMemberData>()

                    val memberIds = members.map { it.user_id }
                    if (memberIds.isNotEmpty()) {
                        val users = memberIds.flatMap { memberId ->
                            try {
                                client.from("users")
                                    .select {
                                        filter {
                                            eq("id", memberId)
                                        }
                                    }
                                    .decodeList<UserData>()
                            } catch (e: Exception) {
                                emptyList()
                            }
                        }

                        chatRoom.setMemberNames(users.mapNotNull { it.account })
                    }

                    // 獲取最後一條訊息
                    val lastMessage = getLastMessage(roomData.id!!)
                    if (lastMessage != null) {
                        chatRoom.setLastMessage(lastMessage.content)
                        chatRoom.setLastMessageTime(lastMessage.created_at)
                    }

                    chatRoom
                }
            } catch (e: Exception) {
                e.printStackTrace()
                emptyList()
            }
        }
    }

    /**
     * 獲取聊天室的最後一條訊息
     */
    private fun getLastMessage(chatroomId: Int): MessageData? {
        return runBlocking {
            try {
                val messages = client.from("messages")
                    .select {
                        filter {
                            eq("chatroom_id", chatroomId)
                        }
                        order("created_at", Order.DESCENDING)
                        limit(1)
                    }
                    .decodeList<MessageData>()

                messages.firstOrNull()
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    /**
     * 發送訊息
     */
    fun sendMessage(chatroomId: Int, senderId: String, content: String): Long? {
        return runBlocking {
            try {
                val message = MessageData(
                    chatroom_id = chatroomId,
                    sender_id = senderId,
                    content = content,
                    created_at = dateFormat.format(Date())
                )

                val createdMessage = client.from("messages")
                    .insert(message) {
                        select(Columns.ALL)
                    }
                    .decodeSingle<MessageData>()

                createdMessage.id?.toLong()
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    /**
     * 獲取聊天室的所有訊息
     */
    fun getMessagesByChatRoom(chatroomId: Int, currentUserId: String): List<Message> {
        return runBlocking {
            try {
                val messages = client.from("messages")
                    .select {
                        filter {
                            eq("chatroom_id", chatroomId)
                        }
                        order("created_at", Order.ASCENDING)
                    }
                    .decodeList<MessageData>()

                // 獲取發送者資訊
                val senderIds = messages.map { it.sender_id }.distinct()
                val users = if (senderIds.isNotEmpty()) {
                    // 查詢所有發送者（簡化版本：逐個查詢）
                    senderIds.flatMap { senderId ->
                        try {
                            client.from("users")
                                .select {
                                    filter {
                                        eq("id", senderId)
                                    }
                                }
                                .decodeList<UserData>()
                        } catch (e: Exception) {
                            emptyList()
                        }
                    }
                } else {
                    emptyList()
                }

                val userMap = users.associateBy { it.id }

                // 轉換為 Message 模型
                messages.map { msgData ->
                    val message = Message(
                        msgData.id ?: 0,
                        msgData.chatroom_id,
                        msgData.sender_id.toIntOrNull() ?: 0,
                        msgData.content,
                        msgData.created_at
                    )

                    val user = userMap[msgData.sender_id]
                    message.setSenderName(user?.account ?: user?.email ?: "未知用戶")
                    message.setSentByMe(msgData.sender_id == currentUserId)

                    message
                }
            } catch (e: Exception) {
                e.printStackTrace()
                emptyList()
            }
        }
    }

    /**
     * 獲取當前用戶 ID
     */
    fun getCurrentUserId(): String? {
        return runBlocking {
            try {
                val session = client.auth.currentSessionOrNull()
                session?.user?.id
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }
}

